### ISSUE TITLE
Default clients list in the Config

### ISSUE THREAD

**Issue by victornoel** [pac4j/pac4j#987]
Currently with pac4j, in the `Config` object, one can define a default `Client` object.

It could/would make sense to be able, instead, to pass a default clients' name list, in the same format that what is passed to the `SecurityLogic` object.

I already implemented it in jax-rs-pac4j at the annotation/filter level because I found it useful, and some people are using it too (see pac4j/jax-rs-pac4j#28) so it could be a desired feature to generalise to pac4j itself.

**Comment by leleuj** [comment]
Yes, it's tracked for version 3.0.0: https://github.com/pac4j/pac4j/issues/938 when rebuilding the callback client mapping: https://github.com/pac4j/pac4j/issues/967

**Comment by victornoel** [comment]
Ok. It is a bit different maybe, since it's about the default client for the `SecurityLogic` and not the callback client, but there is some overlapping yes!

**Comment by leleuj** [comment]
In fact, if the default client is removed from the `Clients` component, it will be mandatory in both filters...

**Comment by victornoel** [comment]
ah ok, I didn't understand that it would be removed from the Config, so we can close this issue, no?

**Comment by leleuj** [comment]
Yes, this is somehow a duplicate of https://github.com/pac4j/pac4j/issues/938

I'm not 100% sure of how you will be able to define a default client at the endpoint level, but this is the plan.

**Comment by victornoel** [comment]
@leleuj in the end, no default client was added to the `DefaultSecurityLogic`/`DefaultSecurityClientFinder`.

Actually, this is how I handle the desired use case in jax-rs-pac4j:
 - I added a `defaultClients` property in `JaxRsConfig` (which extends `Config`). It is `String` that contains client names separated by commas.
 - In my `SecurityFilter` implementation, in case no clients was specified, I use this `defaultClients` information instead.

So for now, I moved that responsibility back to the `Config` in jax-rs-pac4j.

Hence:
 1. I believe this could be a desired feature for all integrations (i.e., specifying the defaults clients to use by the `SecurityLogic`).
 2. As you were planning to do, we can add to that the possibility to define a default client in the `DefaultSecurityClientFinder`, but that's a different use case I think. Also I'm not clear if it should be one default client or a list of default client name in this situation...

What do you think? Should we reopen this?

**Comment by victornoel** [comment]
Actually, to be coherent, I would always use a list of clients names in `Config`, `DefaultSecurityClientFinder` and `DefaultCallbackClientFinder` and resolve them at the last moment.

Why? because this is information meant to be provided by the user via configuration.

**Comment by victornoel** [comment]
Also another problem is that if you want to be able to configure a default client by endpoint, it means that you can't reuse the logics between endpoint since the default client is a configuration of the logic!
And thus the `Config.getXXXLogic()` become useless!

**Comment by leleuj** [comment]
I reopen this issue as I don't the work is complete on this.

**Comment by leleuj** [comment]
In fact, I even missed the point as I "mixed" that with the default client for the callback endpoint.

For the callback endpoint, there are identity providers which don't allow to define a query string in the callback URLs, so we used a default client and removed the client_name parameter from the query string.
It's not that useful in version 3.0.0 as the client_name can be specified via the URL path (and not only via a query parameter).

So we would have a `defaultClients` string in the `Clients` component as well as in the `DefaultCallbackClientFinder` as it's two different use cases?

**Comment by leleuj** [comment]
*Also another problem is that if you want to be able to configure a default client by endpoint, it means that you can't reuse the logics between endpoint since the default client is a configuration of the logic!*

Yes, indeed, it's a limitation, but I don't think it's a big issue.

**Comment by leleuj** [comment]
So just to be sure, what do you propose?

1) a `defaultClient` property in the `DefaultCallbackClientFinder` component (for the callback endpoint)
2) a `defaultClients` property in the `Clients` component used by the `DefaultSecurityLogic` component?

**Comment by victornoel** [comment]
>  a defaultClient property in the DefaultCallbackClientFinder component (for the callback endpoint)

Yes, and even if with the problem with client_name which is now solved, I think it still makes sense to have it…

>  a defaultClients property in the Clients component used by the DefaultSecurityLogic component?

Yes totally :)

**Comment by leleuj** [comment]
OK. That's clear. It makes sense, though I have the feeling it's somehow unbalanced. How about default clients both for the security and callback logics and both at the `Clients` level and at the logic level? But that's maybe too much...

**Comment by millross** [comment]
This does feel like we're creating a lot of defaultClient/defaultClients properties with slightly different names but presumably rather different meanings. Feels like we're going to end up causing all sorts of potential confusion over which to use when. Just my 2c worth. Personally speaking I don't like this kind of approach, where you will get people saying "I've set default client, but I'm not seeing the behaviour I expect" because they should have set something different but similarly named. Documentation won't be the answer to this. At the very least, I'd push for talking about client names rather than clients where we're providing lists of names. But even that - I think - will lead to some confusion.

**Comment by victornoel** [comment]
I agree with you @millross, there is two things for me:
 - a general one: pac4j is like this, it uses a lot of defaults, setters and stuffs to make pac4j customizable. So it's a mess to clearly understand what' happening sometimes, but it's powerfull.
 - this issue: if we consider that the first point is the way we go (and it is the case until now), then we need to at least avoid confusion, so we should maybe choose better names, such as putting `clientName` in the name and not only `client`, but also maybe calling it `defaultSecurityLogicClientNames` (that's ugly :P)?

I think we can go back to the use case: when setting up an application, it makes sense to have a set of default clients used by the security filter.

Do you think it's a valid use case? What would be the alternative solution to this situation?

**Comment by millross** [comment]
I think we should start with the meaning of what the default clients should do in either case, rather than who provides them. And how do they all inter-relate? 

Personally for me, either determining defaults is the responsibility of the config, the client finder, or the caller. I think it's a mistake to attempt to make it the responsibility of all three (or even two of the three).It doesn't feel like we're clear on the responsibility here.

**Comment by victornoel** [comment]
Good idea.
For my use case, this should be the config because to me, the client finder encapsulate a strategy to choose a client, not some configuration.

Also, the caller passes clients, which are not defaults, they are the list of clients concerned by the filter, but if the caller does not pass clients, then the default are used: it does not really concerns the client finder.

**Comment by millross** [comment]
Ok, but we should also keep in mind that responsibilities are generalised, not specific to one use-case. So as other use-cases emerge, we should rethink our responsibility structure. I don't think it's a good idea for an API to distribute that responsibility. Or at least if we want specific variation for use cases, then the naming should reflect when that will apply. But I think it's better to have clearer responsibility.

**Comment by victornoel** [comment]
I'm not sure to understand what you mean ^^ could you explain for this specific question we are currenly discussion what that means?

**Comment by millross** [comment]
If you explain the specific question, then yes. But in principle: responsibilities aren't the same as capabilities. Presumably your use-case isn't the only one, so talking about the responsibility in the context of one use-case isn't sufficient. What about the use-cases which have driven the idea of the defaulting elsewhere? A responsibility which resides over different classes is, in my view, unclear to a consumer of the API. 

I think the discussion should start with "what does default client really mean?" Though overall I tend to agree with you, that I don't think the client finder should operate a default client. I would prefer it to respond with an Optional.empty if not found, and then let the calling code determine what to use if not found via orElse. 

What logics will consume your default client names? Just security logic? If so, and if you expect them to be the same for every call, then rather than being a config property, they should be a property of security logic. If the same for multiple handlers then they should be in the Config, in my view.

**Comment by victornoel** [comment]
> I think the discussion should start with "what does default client really mean?" Though overall I tend to agree with you, that I don't think the client finder should operate a default client. I would prefer it to respond with an Optional.empty if not found, and then let the calling code determine what to use if not found via orElse.

The default client names are not the clients that are used if no clients are found by the client finder, it is the clients used (and passed to the client finder) if the filter does not provide any client names.

> What logics will consume your default client names? Just security logic?

Yes

> If so, and if you expect them to be the same for every call, then rather than being a config property, they should be a property of security logic. If the same for multiple handlers then they should be in the Config, in my view.

That's my point: until now, people using pac4j are passing a list of clients to the security logic (via the filter, which is the embodiement of the logic in a given framework).
When they find themselves with tens of filters and they pass everytime the same set of clients, they start to think: wait, my application is using only those 3 clients, why should I redefine them everytime I define a filter? There should be default list of clients for my application.
Except there ought to be some exceptions, so the security logic needs also to take a client list as a parameter.

We could also say there is a client list for the application and a client list override for the filter (to use a different vocabulary for the same solution).

**Comment by millross** [comment]
"The default client names are not the clients that are used if no clients are found by the client finder, it is the clients used (and passed to the client finder) if the filter does not provide any client names" -> I wasn't suggesting that they were. But the discussion above encompasses the default client property of the client finder too. I'm saying that should be dropped.

If we want to add global stuff for multiple SecurityLogic instances in the Config, I would suggest we need multiple configurations for clarity. I would say that the best model for this would be a global config which offers shared configuration which apply to multiple types of logic, and also contains a SecurityLogicConfig, a CallbackLogicConfig and a LogoutLogic config for configuring just those pieces which we want to apply to all instances of a given logic. 

At the very least this would give us some namespacing on things like "default" stuff and make things hopefully a little clearer to what users of our API are configuring when they set these things.

**Comment by victornoel** [comment]
> "The default client names are not the clients that are used if no clients are found by the client finder, it is the clients used (and passed to the client finder) if the filter does not provide any client names" -> I wasn't suggesting that they were. But the discussion above encompasses the default client property of the client finder too. I'm saying that should be dropped.

For the record, there is no default client on the client finder (for the security logic).
Currently there is only the client list passed to the securitylogic.

> If we want to add global stuff for multiple SecurityLogic instances in the Config, I would suggest we need multiple configurations for clarity. I would say that the best model for this would be a global config which offers shared configuration which apply to multiple types of logic, and also contains a SecurityLogicConfig, a CallbackLogicConfig and a LogoutLogic config for configuring just those pieces which we want to apply to all instances of a given logic.

> At the very least this would give us some namespacing on things like "default" stuff and make things hopefully a little clearer to what users of our API are configuring when they set these things.

It makes sense yes!

**Comment by victornoel** [comment]
> There shouldn’t be a default client on the client finder for callback logic either though. Supplying a default should be the responsibility of its caller. As you suggested above, the finder should be a strategy.

Ah yes, I forgot about the callback, I became focused on the security logic :)

I do think that the caller (i.e., the callback filter) should to be able to specify a client to use, so it should be passed to the CallbackLogic in the same way the client list is passed to the SecurityLogic I suppose… That was the purpose of #938. The conclusion was that since there was the client finder whose responsibility was to choose a client to use based on the url, it was its responsibility to hold the default client.

So:
 - the DefaultCallbackClientFinder is actually a DefaultCallbackClientChooser: it chooses based on the request.
 - the DefaultSecurityClientFinder is a bit different: it chooses, but amongst the list of client given to it.

I wonder also why the DefaultSecurityClientFinder look at the request parameter to do its work, wasn't it something used only for callback?

**Comment by leleuj** [comment]
I'm trying to follow the discussion and all the proposed ideas and arguments.

When I designed the code, I had in mind that the `ClientFinder` is responsible for defining the clients used by the logic.
That's why the default client (for the callback endpoint) can be defined in the `DefaultCallbackClientFinder`:  this is no longer really necessary in version 3 (thanks to the `PathParameterCallbackUrlResolver`), but this was super useful for identity providers which do not support query strings to pass the client name.

*I wonder also why the DefaultSecurityClientFinder look at the request parameter to do its work, wasn't it something used only for callback?*: for the security logic, you can pick up a specific authentication mechanism. Use case: I have a protected URL and I can login with Facebook or Twitter (clients="FacebookClient,TwitterClient"). Calling /protectedURL?client_name=FacebookClient will start a Facebook login process while /protectedURL?client_name=TwitterClient will start a Twitter login process.

I'm in line with the idea that things should be really explicit, so I would go for a very simple solution: I would have a `defaultSecurityClients` and a `defaultCallbackClient` at the `Config` level and then, I would update the `DefaultXClientFinder` to take care of that.

**Comment by victornoel** [comment]
> I wonder also why the DefaultSecurityClientFinder look at the request parameter to do its work, wasn't it something used only for callback?: for the security logic, you can pick up a specific authentication mechanism. Use case: I have a protected URL and I can login with Facebook or Twitter (clients="FacebookClient,TwitterClient"). Calling /protectedURL?client_name=FacebookClient will start a Facebook login process while /protectedURL?client_name=TwitterClient will start a Twitter login process.

Ah yes, totally makes sense :)

> I'm in line with the idea that things should be really explicit, so I would go for a very simple solution: I would have a defaultSecurityClients and a defaultCallbackClient at the Config level and then, I would update the DefaultXClientFinder to take care of that.

Ok, that resolves my original issue, but don't forget the need for specifying a specific client for the callback logic (#938). To be consistent, we could simply do as with the security logic and pass it in the perform method: if it is present, it will be used by the `DefaultCallbackClientFinder`, if it is not then one will be looked up via the url parameters or whatever, and if none is found, the `Config`'s `defaultCallbackClient` would be used.

**Comment by leleuj** [comment]
Once again, given the relative complexity of all use cases, I would stick to a simple rule: "the client finder is responsible for finding the right client".

So I would have that in the `DefaultCallbackClientFinder` with the following algorithm:

1) loop on all `CallbackUrlResolver` to find the right client

2) if none, use the `defaultClient` property (of the `DefaultCallbackClientFinder`)

3) if none, use the `defaultCallbackClient` property of the `Config`.

Sounds good to you?

**Comment by victornoel** [comment]
> Once again, given the relative complexity of all use cases, I would stick to a simple rule: "the client finder is responsible for finding the right client".

In that case why is the `SecurityLogic` taking a list of clients in its perform method? Shouldn't it be the same and the clients to use be passed to the `DefaultSecurityClientFinder`? ;)

**Comment by leleuj** [comment]
The clients passed to the `perform` method are the configured clients of the "security filter/interceptor/handler" specific to the web framework. It's really the `DefaultSecurityClientFinder` which decides if it should use them or keep the one passed in the URL...

`final List<Client> currentClients = clientFinder.find(configClients, context, clients);`

**Comment by victornoel** [comment]
> The clients passed to the perform method are the configured clients of the "security filter/interceptor/handler" specific to the web framework.

Yes, and my point with in #938 was that there was a similar need for the `CallbackLogic`, the client should be chosen in order from:
 - provided by the user via the filter of the specific web framework
 - infered from the url (via `CallbackUrlResolver` currently)
 - use the `defaultCallbackClient` property of the `Config`

I just find that there is in incoherence between they way things are for the security filter and the callback filter, and I wonder why not fix this incoherence once and for all?

**Comment by leleuj** [comment]
In the callback endpoint, the client is taken (by the `DefaultCallbackClientFinder`):

1) from the URL (*client_name* request or path parameter)
2) if blank, from the `defaultClient` configured in the `DefaultCallbackClientFinder`
3) if blank, from the `defaultCallbackClient` in the `Config`class (to be developed).

In the security endpoint, the clients are taken (by the `DefaultSecurityClientFinder`):

1) from the URL (client_name request parameter) and assuming it's in the `clients` definition if not blank or in the `defaultSecurityClients` from the `Config` class (to be developed)
2) if blank, it's the `clients` definition if not blank or the `defaultSecurityClients` from the `Config` class (to be developed)

It seems very similar to me, except that we don't have a `defaultClients` at the `DefaultSecurityClientFinder` level, but we could add it: "the `clients` definition if not blank or the `defaultClients` if not blank or the `defaultSecurityClients` from the `Config` class".

Is this clearer, more consistent to you?

**Comment by leleuj** [comment]
In fact, I realize that the `defaultClient` of the `DefaultCallbackClientFinder` is similar to the "clients" configuration of the security filter.
So maybe we should have a "client"  property at the callback endpoint level. This way, no need to know about the `DefaultCallbackClientFinder`, it would be explicit and similar to the security part.

So either the client(s) is found at the URL level, or at the filter level or at the configuration level, both for the security and callback parts.

Sounds better?

**Comment by victornoel** [comment]
### This is how I see things:
 1. for security filter, the user wants
     a. to specify some clients that should be used for the auth and let pac4j determines if a specific one should be used based on the url or all (they will be tried in order of appearance)
     b. to avoid repetition, the user wants to specify those clients only once at the config level
 2. for callback logic, the user wants
    a. to specify nothing and let pac4j find back the client based on the url
    b. to specify a default client in the config to use if none is found on the url
    c. to force the use of a specific client for a specific endpoint

### Current situation:
**1a** and **2a** are already implemented.
**2b** was implemented (in the `Config`) but it was removed and moved to do the same but by endpoint.
**2b** is useless IMHO: it defeats the purpose of defining a default client if you need to do it for each enpoint.
**1b** is was I am asking for here.
**2c** is what I was asking for in #938.

### Possible places of configuration:
**1a** is about configuring the security filter, so the clients names should be owned by the security filter (and either passed to the `Logic`'s `perfom` operation as currently or configured in the logic, see below)
**2c** is about configuring the callback filter, so the clients names should be owned by the callback filter (and either passed to the `Logic`'s `perfom` operation as currently or configured in the logic, see below)
**2b** should be changed to be simply a default client name stored at the `Config` level
**1b** should be stored as a default clients names at the `Config` level

### Possible algorithm:
**1a** is used first and if not specified, **1b** is used. then the client names on the url can be used to select one but it is out of scope of the current discussion IMHO (by the way there could be multiple clients passed via the url...)
**2c** is used first, if empty **2a** happens and if none found **2b** is used if present.

So security and callback filters does not behave the same and also usages are not the same: most users will rely on **1a**/**1b** and on **2a**.

The `ClientFinder` should be at least responsible of implementing:
 1. for security filter: determines if a specific client should be used based on the url or all
 2. for callback filter: let pac4j find back the client based on the url

To do this there is 3 possibilities:
 1. put everything in the `ClientFinder`, via setters, even **1a** and **2c**, and do not pass the client list to the `perform` operation nor to the `find` operation: so it also takes care of looking into the `Config` for defaults.
 2. only put the algorithm itself in it and passes everything via parameters of `find` and store everything in the `Logic` (via setters), even **1a** and **2c**, and do not pass the clients names list to the `perform` operation, so the logic takes care of looking into the `Config` for defaults.
 3. pass everything to the `perfom` operation.

I like number 2 personally: `ClientFinder` is just a strategy and the `Logic` owns its configuration (there is one logic instance per filter!).
By the way, most of the data passed to the `perfom` operation could actually be owned by the `Logic` and specified via setters IMHO. The `perfom` operation should/could only handle the data contextual to the current request being processed!

Sorry for making things for more complex :P

**Comment by leleuj** [comment]
Wow! I will review that carefully...

**Comment by leleuj** [comment]
I'm in line with what I read and I don't see so many changes. I think we can forget 2b for now.

So we would need to:

1) Add a `defaultSecurityClients` property in the `Config` component
2) Add a `client` input parameter in the `perform` method of the `CallbackLogic`
3) Update the client finders to take that into account.

My main question is whether we should keep the parameters in the `perform`  method or in the logics as properties. It's done this way currently, as I designed the logics to hold algorithms, not configuration (properties).
We have logics defined at the `Config` level: I guess they are used in JAX-RS?

**Comment by victornoel** [comment]
> I'm in line with what I read and I don't see so many changes. I think we can forget 2b for now.

Okay. I would rationalize it like that: there shouldn't be many callback endpoints, so there is no reason to have a default callback client at the `Config` level, it can be configured at the endpoint level (via a `perform` parameter as you propose, or via a `CallbackLogic` configuration as we discuss below).

> My main question is whether we should keep the parameters in the perform method or in the logics as properties. It's done this way currently, as I designed the logics to hold algorithms, not configuration (properties).

This is just some thoughts, but lately I have been trying to improve my object-oriented design and from what I read (I highly recommend the book "Object Thinking") and experienced, a good way of seeing things is the following (sorry, I'm going to write a lot, I'm really into this stuff lately ^^):
 - objects are things with behaviours
  - they should contain everything needed to do their behaviour

With `Logic`s, their behaviour is embedded in `perform`: this behaviour is to perform the security logic for example, and for this there are three things they need as they are currently coded:
   1. some endpoint information (i.e. it is the same thing everytime `perform` is called)
   2. some global application configuration (defaults)
   3. some request-related information (change every time perform is called)

So, for me:
 - when we instantiate ONE `SecurityLogic`, we are at the endpoint level, so we should give it, via the constructor (or setters to accommodate other instantiation means), everything that won't change for this endpoint (including the defaults, via giving giving access to the `Config`).
 - when we ask this object to do its behaviour, we only give it what is needed for it to do it (so the request-related stuffs).

That's a design choice, I like it, but I don't it have to be done like this, especially because we have lots of stuffs depending on pac4j :) it would be great if we did thought, it would simplify integrators work I think (after first complexifying it for the refactor :P) in the long term and ease our thinking with abstractions that are close to reality.

> We have logics defined at the `Config` level: I guess they are used in JAX-RS?

They are exploited by the jax-rs-pac4j integration, but I'm not sure anyone is actually making usage of this possibility. I'm not clear what is the use case for those in fact ^^

**Comment by leleuj** [comment]
I understand your point of view and I opened an issue for this potential improvement: https://github.com/pac4j/pac4j/issues/1049, which is not blocking for this change.

**Comment by victornoel** [comment]
cool, thanks for the vote of confidence ;)
